src/Core/Transform/MutualInformationAffine/Code.py

```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env python
+
+# =========================================================================
+#
+#   Copyright NumFOCUS
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#          http://www.apache.org/licenses/LICENSE-2.0.txt
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# =========================================================================*/
+
+import sys
+import numpy as np
+
+import itk
+
+if len(sys.argv) < 4:
+    raise Exception(f"Usage: {sys.argv[0]} fixed_image moving_image output_image")
+
+# Import images
+fixed_image = itk.imread(sys.argv[1], itk.UC)
+moving_image = itk.imread(sys.argv[2], itk.UC)
+
+# Preprocess images
+fixed_normalized_image = itk.normalize_image_filter(fixed_image)
+fixed_smoothed_image = itk.discrete_gaussian_image_filter(
+    fixed_normalized_image, variance=2.0
+)
+
+moving_normalized_image = itk.normalize_image_filter(moving_image)
+moving_smoothed_image = itk.discrete_gaussian_image_filter(
+    moving_normalized_image, variance=2.0
+)
+
+# Run gradient ascent
+transform = itk.AffineTransform[itk.D, fixed_image.GetImageDimension()].New()
+interpolator = itk.LinearInterpolateImageFunction[type(fixed_image), itk.D].New()
+
+metric = itk.MutualInformationImageToImageMetric[
+    type(fixed_image), type(moving_image)
+].New()
+
+metric.SetNumberOfSpatialSamples(100)
+metric.SetFixedImageStandardDeviation(5.0)
+metric.SetMovingImageStandardDeviation(5.0)
+
+metric.ReinitializeSeed(121212)
+
+optimizer = itk.GradientDescentOptimizer.New()
+
+optimizer.SetLearningRate(1.0)
+optimizer.SetNumberOfIterations(200)
+optimizer.MaximizeOff()
+
+# Set scales so that the optimizer can take
+# large steps along translation parameters,
+# moderate steps along rotational parameters, and
+# small steps along scale parameters
+optimizer.SetScales([100, 0.5, 0.5, 100, 0.001, 0.001])
+
+registrar = itk.ImageRegistrationMethod[type(fixed_image), type(moving_image)].New()
```

Comment:
Can we used the functional Python interface here? e.g.

```
result_transform = itk.image_registration_method(fixed_image=fixed_image, moving_image=moving_image, transform=transform, [...]
```

Reply:
The functional Python interface appears to be expecting images of type `signed short` and fails when `unsigned char` images are provided. I've checked that ITK is wrapped for both types. Is there a way to explicitly set the image type? Is this a wrapping that needs to be added to ITK?

Error:
> Traceback (most recent call last):
  File "Code.py", line 73, in <module>
    result_object = itk.image_registration_method(
  File "C:\repos\ITK-builds-2\Wrapping\Generators\Python\itk\support\itkHelpers.py", line 211, in image_filter_wrapper
    return image_filter(*args, **kwargs)
  File "C:\repos\ITK-builds-2\Wrapping\Generators\Python\itk\itkImageRegistrationMethodPython.py", line 581, in image_registration_method
    instance = itk.ImageRegistrationMethod.New(*args, **kwargs)
  File "C:\repos\ITK-builds-2\Wrapping\Generators\Python\itk\support\itkTemplate.py", line 726, in New
    return self[list(keys)[0]].New(*args, **kwargs)
  File "C:\repos\ITK-builds-2\Wrapping\Generators\Python\itk\itkImageRegistrationMethodPython.py", line 339, in New
    itkTemplate.New(obj, *args, **kargs)
  File "C:\repos\ITK-builds-2\Wrapping\Generators\Python\itk\support\itkTemplate.py", line 792, in New
    itk.set_inputs(self, args, kargs)
  File "C:\repos\ITK-builds-2\Wrapping\Generators\Python\itk\support\itkExtras.py", line 928, in set_inputs
    attrib(itk.output(value))
TypeError: in method 'itkImageRegistrationMethodISS2ISS2_SetFixedImage', argument 2 of type 'itkImageSS2 const *'

Verified input types:
> \> print(type(fixed_image))
<class 'itk.itkImagePython.itkImageUC2'>
\> print(type(moving_image))
<class 'itk.itkImagePython.itkImageUC2'>

Reply:
Also note that `itk.image_registration_method` returns an object of type `'itk.itkTransformBasePython.itkDataObjectDecoratorTD22'` which cannot be passed directly to `resample_image_filter` as an `itk.transform`. Not a big deal given that the registration transform is updated in-place and can be passed on, but I'm not sure what this object contains. Where can I find documentation for the functional interface?

Reply:
We can be explicit with the types we want with the [`ttype` keyword argument](https://github.com/InsightSoftwareConsortium/ITK/blob/2880b97c88100ddaf809204d6a156bfb5e694e85/Wrapping/Generators/Python/itk/support/itkTemplate.py#L678).

However, we could have the filter select the proper default template type by looking at the first positional argument / template argument. This corresponds to the first indexed input to the filter, the primary input. While `itk::ImageRegistrationMethod` does not identify the `FixedImage` as its primary input, `itk::ImageRegistrationMethodv4` does:

https://github.com/InsightSoftwareConsortium/ITK/blob/cde72b3e7d716542b11d2dbb402bbe5816cf731f/Modules/Registration/RegistrationMethodsv4/include/itkImageRegistrationMethodv4.hxx#L45

So, we should update this examples to use the v4 registration method.

Reply:
This example currently relies on the Viola-Wells mutual information metric but it seems that only the Mattes mutual information metric has a v4 counterpart. Given that a v4 metric is required for the v4 registration method, is it suitable to explicitly set `ttype` in this example for the v3 metric and open an issue to add a new example using the Mattes mutual information v4 metric? It doesn't look like we currently have an example making use of `MattesMutualInformationImageToImageMetricv4`.